Add new helper for matching reauth/reconfigure config flows

epenet opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
When using `reauth`/`reconfigure`, there is currently no easy way to check that the `unique_id` of the original entry still matches what the new `unique_id` would be with the latest input.
This is especially obvious when `reauth`/`reconfigure` flows reuse existing functionnality from `user` flow.

I propose that we add this helper to make it easier to bail out if there is a mismatch:

Sample use:
```python
await self.async_set_unique_id(uid)
if self.source == SOURCE_REAUTH:
    self._abort_if_unique_id_mismatch(reason="reauth_account_mismatch")
    return self.async_update_reload_and_abort(self._get_reauth_entry(), data=data)
self._abort_if_unique_id_configured()
return self.async_create_entry(title=uid, data=data)
```

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/developers.home-assistant/pull/2353

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

epenet committed:
Add new helper for matching reauth/reconfigure config flows

epenet committed:
Simplify

epenet committed:
error -> reason

epenet committed:
Unrelated

epenet committed:
Remove guard

epenet committed:
Add tests

epenet committed:
Remove invalid comment

frenck commented on homeassistant/config_entries.py:
```diff
@@ -2432,6 +2432,26 @@ def _async_abort_entries_match(
+        in user visible flows.
+        """
+        if (
+            self.source == SOURCE_REAUTH
+            and self._get_reauth_entry().unique_id != self.unique_id
+        ) or (
+            self.source == SOURCE_RECONFIGURE
+            and self._get_reconfigure_entry().unique_id != self.unique_id
```
I'm not sure if we should assume the unique ID is being used for detecting these cases.

frenck commented:
> I think it makes sense to add this helper.

Can you elaborate on that? There is nothing in the PR description that explain the foundation behind it.

epenet commented:
> > I think it makes sense to add this helper.
> 
> Can you elaborate on that? There is nothing in the PR description that explain the foundation behind it.

It's a pattern I've bumped into a few times. Axis and abode for example use this pattern.
It was used by integrations when reconfigure wasn't available.
I'll look for more examples on Monday

epenet commented:
> > > I think it makes sense to add this helper.
> > 
> > 
> > Can you elaborate on that? There is nothing in the PR description that explain the foundation behind it.
> 
> It's a pattern I've bumped into a few times. Axis and abode for example use this pattern. It was used by integrations when reconfigure wasn't available. I'll look for more examples on Monday

It's also linked to #127527

epenet commented:
> I'll look for more examples on Monday

Here's another example I've just bumped into, which I think would benefit from this:
https://github.com/home-assistant/core/pull/127726/files#r1789215553

epenet committed:
Apply to bmw_connected_drive

epenet committed:
Apply to tesla_fleet

epenet commented:
> > I think it makes sense to add this helper.
> 
> Can you elaborate on that? There is nothing in the PR description that explain the foundation behind it.

- I have adjusted the PR description
- I have added a documentation PR
- I have added tests
- I have applied the helper to three integrations that already have some of this logic implemented, to show the benefit
  - bmw_connected_drive
  - spotify
  - tesla_fleet

frenck reviewed (commented):
Honestly, I find the use-case/gain to shallow to add a helper. While helpers can be helpful, they can also make code harder to read / harder to follow (as one has to jump between more logic).

I'm not sure on this one. Going to ask for a second opinion from another reviewer.

../Frenck

epenet commented:
> Honestly, I find the use-case/gain to shallow to add a helper. While helpers can be helpful, they can also make code harder to read / harder to follow (as one has to jump between more logic).
> 
> I'm not sure on this one. Going to ask for a second opinion from another reviewer.
> 
> ../Frenck

One more note: it is a mirror to `self._abort_if_unique_id_configured` that cannot be used inside REAUTH/RECONFIGURE flows.

MartinHjelmare commented:
I think it's nice since it will help integrations to do the right thing. I think there are many integrations that have reauth flows where they don't check that the flow is reauthenticating matching flow data.

frenck reviewed (approved):
Thanks, @epenet 👍 

../Frenck
